### ISSUE TITLE
EditorVR Contexts

### ISSUE THREAD

**Issue by amirebrahimi** [Unity-Technologies/EditorVR#51]
Originally when we were discussing EditorVR design we knew that we'd want to experiment with every part of the system from menus, UIs, tools, etc. One idea we had at the time is that a group of these might be a complete package from a developer. So, the developer might have the need to re-skin all of EditorVR. This isn't entirely what @leith-bartrich is alluding to in #48, but it has parts of it.

So, details need to be flushed out about what a context encompasses. I suppose without a context that you'd be in VR staring at an empty scene with possibly controllers, but nothing else -- no menus, no tools, no workspaces, etc. It begs the question, what is the UX for switching contexts?

Aside from that, there would still need to be a default context at startup as some people are coming to EditorVR as a user with not much Unity knowledge and not as a developer who is looking to build tools or a custom VR editing environment. So, it would be a non-starter if a user had to code something or wire up some settings in a ScriptableObject on disk or a menu somewhere.

Opening this up for comments.

**Comment by leith-bartrich** [comment]
Agree with above.  I've previously built a system like this for touch based content creation where screen space was limited, but complex content creation was what needed to be done.  The idea was a context stack.  Where the top most context was the one "current" one.  Those below it on the stack were "subverted."   They maintained state but took no action.  They got out of the way.  The current context has exclusive control over UI and calls to the system.  It can also create another "sub-context" and put it on the top of the stack (push, which would subvert the current context.)  It can also exit (pop), which removes it from the top of the stack.  The "subverted" context below it becomes current one and takes back control.

The bottom most context was the "root" context.  Which in my case, was a top level menu system in which you choose what you are trying to do.  A call to exit the "root" context is invalid, or simply ignored.  Though in Unity's case, I would guess that closing the VRView windows closes out the whole system and does technically exit or pop the root context.  Though that's being pedantic.

If you followed this methodology, you'd have the "root" context on by default.  So there would be no need for users to wire it up.  Its job would be to enumerate all available contexts, make them available as easy to execute.  And possibly some functionality to pick out groups of contexts and put them in a quick access panel.  Kind of like arranging shortcuts on the desktop, or organizing your start menu. In fact, its really just a shell.  Much like SteamVR is a shell.  Or the start menu is a shell.  Or the old ms-dos prompt is a shell.  Or Norton Commander... bash...  etc.

Your existing scene editing system, with all your "tools" and "workspaces" and such, could be wrapped up into a context in that menu, called "SceneEditor" or "Unity Scene Editor."  Nice and big and easy to point at and enter right away.

Without a particular UI design in mind for the root context, I'd suggest you consider stealing (paying great respect to) SteamVR's concept as a possible design.  At least to start.  They have a big wall sized screen that you use to organize and run applications.  You could easily mimic exactly that concept and organize and enter contexts rather than applications.

That same context enumeration and picking system might be something you expose as a library too, such that someone making their own context, could enumerate and provide their own "sub-contexts" in the same familiar manner.  Or, they might choose to enter sub-contexts via their own methodology.

That's one idea anyway.

I'd point out that it probably should respond to keyboard and mouse commands as well as VR controllers.  To support the possibility of desk-bound and seated VR.  Sub-contexts may not support keyboard and mouse at all.  But the root context probably should expect to need to.

While it's not necessary to define this, a good convention is probably to borrow SteamVR's menu button to pop contexts where reasonable.  I believe the OpenVR API lets you intercept that button and probably, it should be used to pop contexts.  Unless you're in the "root" context, in which case it should probably be returned to the OpenVR SDK's control, which would result in it popping you back out to the SteamVR menu I believe.  Contexts should be allowed to exit (pop) via other programmatic means as they deem appropriate, as well.

**Comment by amirebrahimi** [comment]
Adding a few more thoughts from a conversation with @mtschoen. Potential first step:

```
interface IContext
{
    public void Setup()
    public void Teardown()
}

class EditorVRContext : ScriptableObject, IContext
{
    [SerializeField]
    List<Type> defaultTools;
    // Any other settings that would make sense to EditorVR
}

class AlternateVRContext : ScriptableObject, IContext
{
}
```

Edit->Preferences... would have a new menu for setting the context that loads for EditorVR by default.
A ContextSelector is what attaches to VRView and on initialization, spawns the context that was selected.

**Comment by amirebrahimi** [comment]
Merged #58 into custom branch features/amir/contexts @ 59fc85feeda67e42abc89c47c7486c8bad76d135
